### PR TITLE
Min time between induced gcs

### LINKED ISSUE #7431: Recent memory optimizations made archive sync significantly slower
After optimizing memory usage on archive it seems like archive is much much slower than it used to be on 1.28.0.
Below screenshot for gnosis which is common use case.

image

### PR DESCRIPTION
**Pull request by benaadams**
Resolves #7431

## Changes

- Add min time between forced GCs
- Add periodic timer to GC during idle
- GC on background thread

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
Can we move this code to separate component like we have `GCKeeper`?

**Comment by benaadams** [comment]
> Can we move this code to separate component like we have `GCKeeper`?

Moved to separate type; also added some coordination with the GCKeeper

**Comment by benaadams** [comment]
Hmm broke something in the extraction; is calling GC all the time now 🤔

File: src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs [hunk 1]
```diff
@@ -165,9 +167,7 @@ private async Task ScheduleGCInternal()
                     GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
                 }
 
-                System.GC.Collect((int)generation, mode, blocking: true, compacting: compacting > 0);
-
-                MallocHelper.Instance.MallocTrim((uint)1.MiB());
+                GCScheduler.Instance.GCCollect((int)generation, mode, blocking: true, compacting: compacting > 0);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs:170]: should lines [165-168](https://github.com/NethermindEth/nethermind/pull/7432/files#diff-79d27af20a7736bf17e78b4578c54be208f0753194a14cfa53557ae2f9a6153fR165-R168) go into this method in GCScheduler?

File: src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs [hunk 2]
```diff
@@ -29,7 +30,7 @@ public GCKeeper(IGCStrategy gcStrategy, ILogManager logManager)
     public IDisposable TryStartNoGCRegion(long? size = null)
     {
         size ??= _defaultSize;
-        if (_gcStrategy.CanStartNoGCRegion())
+        if (GCScheduler.MarkGCPaused() && _gcStrategy.CanStartNoGCRegion())
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs:33]: So if there is a GC already we won't start NoGCRegion, right?

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        return Interlocked.CompareExchange(ref _isPerformingGC, 1, 0) == 0;
+    }
+
+    /// <summary>
+    /// Marks that garbage collection has finished.
+    /// </summary>
+    public static void MarkGCResumed()
+    {
+        Volatile.Write(ref _isPerformingGC, 0);
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs:129]: So instead `_isPerformingGC` we should have `_canPerformGC` with `0` being `true`?

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+namespace Nethermind.Consensus;
+
+public sealed class GCScheduler
+{
+    // Thresholds and limits for triggering garbage collection
+    private const int BlocksBacklogTriggeringManualGC = 4;
+    private const int MaxBlocksWithoutGC = 250;
+    private const int MinSecondsBetweenForcedGC = 120;
+
+    // Flag indicating if a garbage collection is currently in progress
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs:19]: or is disallowed

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            _countToGC--;
+            // Trigger GC if block limit reached and minimum time elapsed
+            if (_countToGC <= 0 && _stopwatch.Elapsed.TotalSeconds > MinSecondsBetweenForcedGC)
+            {
+                _fireGC = false;
+                _stopwatch.Reset();
+                // Ensure only one GC task runs at a time
+                if (_lastGcTask.IsCompleted)
+                {
+                    _lastGcTask = PerformFullGCAsync();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs:58]: Should this be interlocked or something?

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+        // Avoid setting the timer if there are items in the queue or timer is already set
+        if (queueCount > 0 || _gcTimerSet)
+        {
+            // Don't switch on if still processing blocks
+            return;
+        }
+        _gcTimerSet = true;
+        // Schedule GC every 2 minutes when idle
+        _gcTimer.Change(TimeSpan.FromMinutes(2), TimeSpan.FromMinutes(2));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs:71]: Minor/me: I dislike `if return` pattern unless on top of the method, would prefer nesting.

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs (hunk 5 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs:58]: Currently is all called on same RunProcessingLoop thread

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs]: .NET 9 allows enums to be used with `CompareExchange` so we can clean up some of these unhelpful 0, 1

Renamed and added some constants

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs (hunk 4 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs]: Added

File: src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs (hunk 6 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs]: Inverted the ifs

File: src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs]: Wasn't my intention. Instead am now capturing the return and if it was true unpause in Dispose, but don't prevent NoGCRegion

**Pull request merged**